### PR TITLE
Configurable PFCP Heartbeat parameters

### PR DESCRIPTION
**Pull request by sergeymatov**
Use binapi or CLI to configure PFCP Heartbeat interval (timeout)
and number of retries to be sent before association drop.

### CODE REVIEW

File: upf/upf.c [hunk 1]
```diff
@@ -380,6 +380,23 @@ vnet_upf_node_id_set (const pfcp_node_id_t * node_id)
...
+  pfcp_server_main_t *psm = &pfcp_server_main;
+
+  if ((timeout > 60) || (retries > 10))
+    return 1;
+
+  if (timeout)
+    psm->hb_cfg.timeout = timeout;
+
+  if (retries)
+    psm->hb_cfg.retries = retries;
```
Comment (ivan4th) [inline upf/upf.c:395]: What if I want to set `0` retries for some kind of test?
Wouldn't `~0` make more sense as the "don't update" value?

File: upf/upf_cli.c [hunk 2]
```diff
@@ -1646,6 +1646,67 @@ VLIB_CLI_COMMAND (upf_add_policy_command, static) =
...
+  if (rv)
+    error =
+      clib_retur_error
+      ("Invalid parameters. Correct ranges:\n 0 < timeout < 60; 0 < retries < 10");
+  return error;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (upf_pfcp_heartbeats_command, static) = {
+    .path = "upf pfcp hearbeats",
```
Comment (ivan4th) [inline upf/upf_cli.c:1683]: Typo: `hearbeats`

File: upf/upf_api.c [hunk 3]
```diff
@@ -845,6 +845,54 @@ vl_api_upf_pfcp_policer_show_t_handler (vl_api_upf_pfcp_policer_show_t * mp)
...
+
+  rmp = vl_msg_api_alloc (sizeof (*rmp));
+  clib_memset (rmp, 0, sizeof (*rmp));
+
+  rmp->_vl_msg_id =
+    htons (VL_API_UPF_PFCP_HEARTBEATS_SHOW_REPLY + sm->msg_id_base);
+  rmp->context = mp->context;
+
+  rmp->timeout = htonl (psm->hb_cfg.timeout);
+  rmp->retries = htonl (psm->hb_cfg.retries);
```
Comment (ivan4th) [inline upf/upf_api.c:890]: Why not `clib_host_to_net_u32` for the sake of consistency?

File: upf/upf_api.c [hunk 4]
```diff
@@ -845,6 +845,54 @@ vl_api_upf_pfcp_policer_show_t_handler (vl_api_upf_pfcp_policer_show_t * mp)
...
+					   mp)
+{
+  vl_api_upf_pfcp_heartbeats_show_reply_t *rmp = NULL;
+  upf_main_t *sm = &upf_main;
+  pfcp_server_main_t *psm = &pfcp_server_main;
+  vl_api_registration_t *reg;
+
+  reg = vl_api_client_index_to_registration (mp->client_index);
+  if (!reg)
+    {
```
Comment (ivan4th) [inline upf/upf_api.c:878]: `{` and `}` not needed for the single statement

File: upf/upf_cli.c (hunk 2 above)
Comment (ivan4th) [reply upf/upf_cli.c:1683]: Also wouldn't it be better be called `upf pfcp heartbeat-config`?

**Review by ivan4th — CHANGES_REQUESTED** [review]
It would be really great to have some simple & basic verification in E2E as part of this PR, especially given that there's a typo in CLI command path that makes it non-working. Just smth as simple as setting the values through the API and then querying them via CLI, and then setting the values via CLI and querying them via API.

I'd consider it a possibility for a follow-up, but the history of our previously added binapis is rather unambiguous, no new APIs work till they're verified in E2E, and writing E2E tests almost always resulted in fixing the APIs and/or CLIs.

File: upf/upf_cli.c [hunk 5]
```diff
@@ -1646,6 +1646,67 @@ VLIB_CLI_COMMAND (upf_add_policy_command, static) =
...
+  pfcp_server_main_t *psm = &pfcp_server_main;
+  vlib_cli_output (vm, "Timeout: %d Retries: %d", psm->hb_cfg.timeout,
+		   psm->hb_cfg.retries);
+  return NULL;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (upf_show_pfcp_heartbeats_command, static) =
+{
+  .path = "show upf heartbeats",
```
Comment (ivan4th) [inline upf/upf_cli.c:1703]: that should also be probably `show upf heartbeat-config`, as this is indeed heartbeat configuration, not heartbeats themselves

File: upf/upf.api [hunk 6]
```diff
@@ -324,3 +324,22 @@ define upf_pfcp_policer_show_reply {
...
+  u32 client_index;
+  u32 context;
+};
+
+define upf_pfcp_heartbeats_show_reply {
+  u32 client_index;
+  u32 context;
+  u32 timeout;
+  u32 retries;
+};
```
Comment (ivan4th) [inline upf/upf.api:345]: Although we have other `_show` functions in our API already, I don't really like this convention, as APIs don't "show" anything (don't print anything to the console, unlike CLIs), they just return some info.
In this particular case, I'd name the APIs `upf_pfcp_set_heartbeat_config` and `upf_pfcp_get_heartbeat_config`

File: upf/upf.c [hunk 7]
```diff
@@ -380,6 +380,23 @@ vnet_upf_node_id_set (const pfcp_node_id_t * node_id)
   return VNET_API_ERROR_INVALID_ARGUMENT;
 }
 
+int
+vnet_upf_pfcp_heartbeat_config (u32 timeout, u32 retries)
+{
+  pfcp_server_main_t *psm = &pfcp_server_main;
+
+  if ((timeout > 60) || (retries > 10))
```
Comment (ivan4th) [inline upf/upf.c:388]: redundant parens '(...)`

File: upf/upf_api.c [hunk 8]
```diff
@@ -845,6 +845,54 @@ vl_api_upf_pfcp_policer_show_t_handler (vl_api_upf_pfcp_policer_show_t * mp)
   vl_api_send_msg (reg, (u8 *) rmp);
 }
 
+/* API message handler */
+static void
+vl_api_upf_pfcp_heartbeats_set_t_handler (vl_api_upf_pfcp_heartbeats_set_t *
+					  mp)
+{
+  vl_api_upf_pfcp_heartbeats_set_reply_t *rmp = NULL;
+  u32 retries, timeout = 0;
```
Comment (ivan4th) [inline upf/upf_api.c:854]: What purpose does `timeout = 0` assignment serve here?

File: upf/upf_cli.c [hunk 9]
```diff
@@ -1646,6 +1646,67 @@ VLIB_CLI_COMMAND (upf_add_policy_command, static) =
...
+};
+/* *INDENT-ON* */
+
+static clib_error_t *
+upf_show_pfcp_heartbeats_command_fn (vlib_main_t * vm,
+				     unformat_input_t * main_input,
+				     vlib_cli_command_t * cmd)
+{
+  pfcp_server_main_t *psm = &pfcp_server_main;
+  vlib_cli_output (vm, "Timeout: %d Retries: %d", psm->hb_cfg.timeout,
```
Comment (ivan4th) [inline upf/upf_cli.c:1695]: Unsigned values are formatted via `%u` instead of `%d`

File: upf/upf_cli.c [hunk 10]
```diff
@@ -1646,6 +1646,67 @@ VLIB_CLI_COMMAND (upf_add_policy_command, static) =
...
+      else
+	return (clib_error_return (0, "unknown input '%U'",
+				   format_unformat_error, line_input));
+    }
+
+  rv = vnet_upf_pfcp_heartbeat_config (timeout, retries);
+  if (rv)
+    error =
+      clib_retur_error
+      ("Invalid parameters. Correct ranges:\n 0 < timeout < 60; 0 < retries < 10");
```
Comment (ivan4th) [inline upf/upf_cli.c:1677]: This is not consistent with the range check that verifies that `timeout > 60 || retries > 10` is false.
Also, wouldn't it make sense to have max timeout and retries as constants? Also, why have such a low max retries and timeout, at all? (maybe some "meaningless" values are needed for testing)

File: upf/upf_cli.c [hunk 11]
```diff
@@ -1646,6 +1646,67 @@ VLIB_CLI_COMMAND (upf_add_policy_command, static) =
...
+      else if (unformat (line_input, "retries %u", &retries));
+      else
+	return (clib_error_return (0, "unknown input '%U'",
+				   format_unformat_error, line_input));
+    }
+
+  rv = vnet_upf_pfcp_heartbeat_config (timeout, retries);
+  if (rv)
+    error =
+      clib_retur_error
```
Comment (ivan4th) [inline upf/upf_cli.c:1676]: Typo: `clib_retur_error`. This will fail during runtime with an undefined symbol error

File: upf/upf_api.c (hunk 4 above)
Comment (ivan4th) [reply upf/upf_api.c:878]: it's copied over from other places, so perhaps can leave it as-is, or correct everywhere

File: upf/upf_api.c (hunk 3 above)
Comment (sergeymatov) [reply upf/upf_api.c]: fixed

File: upf/upf_cli.c (hunk 2 above)
Comment (sergeymatov) [reply upf/upf_cli.c]: yep, fixed typo and naming

File: upf/upf.c (hunk 1 above)
Comment (sergeymatov) [reply upf/upf.c]: yep, agree. Fixed

File: upf/upf_cli.c (hunk 5 above)
Comment (sergeymatov) [reply upf/upf_cli.c]: renamed

File: upf/upf_pfcp_server.c [hunk 12]
```diff
@@ -402,7 +402,13 @@ request_t1_expired (u32 seq_no)
 static void
 upf_pfcp_server_send_request (pfcp_msg_t * msg)
 {
-  enqueue_request (msg, 3, 10);
+  pfcp_server_main_t *psm = &pfcp_server_main;
+
+  if (pfcp_msg_type (msg->data) == PFCP_HEARTBEAT_REQUEST)
+    enqueue_request (msg, psm->hb_cfg.retries, psm->hb_cfg.timeout);
+  else
+    enqueue_request (msg, 3, 10);
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf_pfcp_server.c]: Done as default request parameters

File: upf/upf_pfcp_server.c [hunk 13]
```diff
@@ -1421,6 +1427,10 @@ clib_error_t *pfcp_server_main_init (vlib_main_t * vm)
 
   upf_debug ("PFCP: start_time: %p, %d, %x.", psm, psm->start_time,
 	     psm->start_time);
+
+  psm->hb_cfg.retries = PFCP_HB_RETRIES;
+  psm->hb_cfg.timeout = PFCP_HB_INTERVAL;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf_pfcp_server.c]: renamed

File: upf/upf.c (hunk 7 above)
Comment (sergeymatov) [reply upf/upf.c]: fixed

File: upf/upf_api.c (hunk 8 above)
Comment (sergeymatov) [reply upf/upf_api.c]: changed to `~0`

File: upf/upf_cli.c (hunk 9 above)
Comment (sergeymatov) [reply upf/upf_cli.c]: fixed

File: upf/upf.api (hunk 6 above)
Comment (sergeymatov) [reply upf/upf.api:345]: @ivan4th thanks for noticing about e2e test case
added as a separate commit

File: test/e2e/upg_e2e.go [hunk 14]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
 			gomega.Expect(showReply.FifoSize).To(gomega.BeEquivalentTo(512)) // KB
 			gomega.Expect(showReply.PreallocFifos).To(gomega.BeEquivalentTo(0))
 		})
+		ginkgo.It("Configures PFCP Heartbeats", func() {
+
```
Comment (ivan4th) [inline test/e2e/upg_e2e.go:734]: Remove the redundant blank line

**Review by ivan4th — CHANGES_REQUESTED** [review]
Some minor quirks

File: upf/upf.c [hunk 15]
```diff
@@ -380,6 +380,22 @@ vnet_upf_node_id_set (const pfcp_node_id_t * node_id)
...
 }
 
+int
+vnet_upf_pfcp_heartbeat_config (u32 timeout, u32 retries)
+{
+  pfcp_server_main_t *psm = &pfcp_server_main;
+
+  if (!timeout || timeout > PFCP_MAX_HB_INTERVAL
+      || retries > PFCP_MAX_HB_RETRIES)
+    return 1;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf.c]: done

File: upf/upf.c [hunk 16]
```diff
@@ -380,6 +380,22 @@ vnet_upf_node_id_set (const pfcp_node_id_t * node_id)
...
+
+  psm->hb_cfg.timeout = timeout;
+
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf.c]: removed

File: test/e2e/upg_e2e.go [hunk 17]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
...
+			reply := &upf.UpfPfcpHeartbeatsSetReply{}
+			err := f.VPP.ApiChannel.SendRequest(hbConfig).ReceiveReply(reply)
+			gomega.Expect(err).To(gomega.BeNil())
+
+			hbConfigStr, er := f.VPP.Ctl("show upf heartbeat-config")
+			gomega.Expect(er).To(gomega.BeNil())
+			gomega.Expect(hbConfigStr).To(gomega.ContainSubstring("Timeout: 10"))
+			gomega.Expect(hbConfigStr).To(gomega.ContainSubstring("Retries: 5"))
+
+			_, er = f.VPP.Ctl("upf pfcp heartbeat-config timeout 5 retries 15")
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

File: test/e2e/upg_e2e.go [hunk 18]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
...
+
+			hbConfig := &upf.UpfPfcpHeartbeatsSet{
+				Retries: 5,
+				Timeout: 10,
+			}
+			reply := &upf.UpfPfcpHeartbeatsSetReply{}
+			err := f.VPP.ApiChannel.SendRequest(hbConfig).ReceiveReply(reply)
+			gomega.Expect(err).To(gomega.BeNil())
+
+			hbConfigStr, er := f.VPP.Ctl("show upf heartbeat-config")
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

File: test/e2e/upg_e2e.go [hunk 19]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
...
+
+			_, er = f.VPP.Ctl("upf pfcp heartbeat-config timeout 5 retries 15")
+			gomega.Expect(er).To(gomega.BeNil())
+			hbGetRequest := &upf.UpfPfcpHeartbeatsGet{}
+			hbGetReply := &upf.UpfPfcpHeartbeatsGetReply{}
+			err =  f.VPP.ApiChannel.SendRequest(hbGetRequest).ReceiveReply(hbGetReply)
+			gomega.Expect(err).To(gomega.BeNil())
+			gomega.Expect(hbGetReply.Timeout).To(gomega.BeEquivalentTo(5))
+			gomega.Expect(hbGetReply.Retries).To(gomega.BeEquivalentTo(15))
+
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: removed

File: test/e2e/upg_e2e.go [hunk 20]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
...
+			gomega.Expect(hbConfigStr).To(gomega.ContainSubstring("Timeout: 10"))
+			gomega.Expect(hbConfigStr).To(gomega.ContainSubstring("Retries: 5"))
+
+			_, er = f.VPP.Ctl("upf pfcp heartbeat-config timeout 5 retries 15")
+			gomega.Expect(er).To(gomega.BeNil())
+			hbGetRequest := &upf.UpfPfcpHeartbeatsGet{}
+			hbGetReply := &upf.UpfPfcpHeartbeatsGetReply{}
+			err =  f.VPP.ApiChannel.SendRequest(hbGetRequest).ReceiveReply(hbGetReply)
+			gomega.Expect(err).To(gomega.BeNil())
+			gomega.Expect(hbGetReply.Timeout).To(gomega.BeEquivalentTo(5))
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

File: test/e2e/upg_e2e.go [hunk 21]
```diff
@@ -730,6 +730,31 @@ var _ = ginkgo.Describe("UPG Binary API", func() {
 			gomega.Expect(showReply.FifoSize).To(gomega.BeEquivalentTo(512)) // KB
 			gomega.Expect(showReply.PreallocFifos).To(gomega.BeEquivalentTo(0))
 		})
+		ginkgo.It("Configures PFCP Heartbeats", func() {
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

**Review by ivan4th — APPROVED** [review]
LGTM

**Pull request merged**